query-algebrizer/src/clauses/mod.rs

```diff
@@ -355,6 +334,73 @@ impl ConjoiningClauses {
         }
     }
 
+    /// Like `constrain_var_to_type` but in reverse: this expands the set of types
+    /// with which a variable is associated.
+    fn broaden_types(&mut self, additional_types: BTreeMap<Variable, HashSet<ValueType>>) {
+        for (var, new_types) in additional_types {
+            match self.known_types.entry(var) {
+                Entry::Vacant(e) => {
+                    if new_types.len() == 1 {
+                        self.extracted_types.remove(e.key());
+                    }
+                    e.insert(new_types);
+                },
+                Entry::Occupied(mut e) => {
+                    e.get_mut().extend(new_types.into_iter());
+                },
+            }
+        }
+    }
+
+    fn narrow_types_for_var(&mut self, var: Variable, types: HashSet<ValueType>) {
```

Comment:
`narrow_types` doesn't set the types — it intersects them with the types that are already there. `broaden` broadens, `narrow` narrows.

```
                let intersected: HashSet<_> = types.intersection(e.get()).cloned().collect();
```